addons/mail/static/src/js/wysiwyg.js

```diff
@@ -0,0 +1,97 @@
+                position: `bottom-start`,
+                onClose: () => {
+                    this.focus();
+                },
+            });
+        }
+    },
+    async startEdition() {
+        const res = await super.startEdition(...arguments);
+        // Only enable mention list in full chatter mode web editor
+        if (this.options.recordInfo?.res_model === "mail.compose.message") {
+            this.odooEditor.document.addEventListener(
+                "keydown",
+                this.triggerMentionList.bind(this),
+                true
+            );
+        }
+        return res;
+    },
+    destroy() {
+        super.destroy();
+        if (this.options.recordInfo?.res_model === "mail.compose.message" && this.odooEditor) {
+            this.odooEditor.document.removeEventListener(
+                "keydown",
+                this.triggerMentionList.bind(this),
+                true
+            );
+        }
+    },
+    triggerMentionList(ev) {
+        const selection = document.getSelection();
+        if (
+            this.isSelectionInEditable() &&
+            selection.isCollapsed &&
+            selection.rangeCount &&
+            !this.mentionList.isOpen &&
+            (ev.key === "@" || ev.key === "#")
+        ) {
+            this.stepBeforeMention = this.odooEditor._historySteps.length - 1;
+            const closest = closestBlock(this.odooEditor.document.getSelection().anchorNode);
+            this.mentionList.open(closest, {
+                type: ev.key === "@" ? "partner" : "channel",
+                onSelect: this.selectMention.bind(this),
+            });
+        }
+    },
+    selectMention: function (ev, option) {
+        this.mentionList.close();
+        const attrs = {};
+        if (option.partner) {
+            Object.assign(attrs, {
+                placeholder: "@" + option.label,
+                model: "res.partner",
+                class: "o_mail_redirect",
+                id: option.partner.id,
+            });
+        } else {
+            Object.assign(attrs, {
+                placeholder: "#" + option.label,
+                model: "discuss.channel",
+                class: "o_channel_redirect",
+                id: option.channel.id,
+            });
+        }
+
+        const mentionBlock = renderToElement("mail.wysiwyg_mention_link", {
+            href: `${url("/web")}#model=${attrs.model}&id=${attrs.id}`,
+            attClass: attrs.class,
+            dataOeId: attrs.id,
+            dataOeModel: attrs.model,
+        });
+
+        const nameNode = document.createTextNode(attrs.placeholder);
+        const space = document.createTextNode("\u00A0");
+        mentionBlock.appendChild(nameNode);
+        this.odooEditor.historyRevertUntil(this.stepBeforeMention);
+        this.odooEditor.execCommand("insert", mentionBlock);
+        this.odooEditor.execCommand("insert", space);
+        setCursorEnd(space, false);
+        this.odooEditor.historyStep();
```

Comment:
Content can be edited while the mention list stays open.
The current strategy to use history before mention doesn't work when editing the content while it's open:
![Sep-14-2023 11-40-15](https://github.com/odoo/odoo/assets/6569390/09536977-5a35-4cf7-bd3c-a1189e380d34)

We could keep same strategy but reduce the lifespan of the mention list. Whenever the content of web editor is changed, we should re-trigger the mention list.